proposals/0000-no-incomplete.rst

```diff
@@ -0,0 +1,268 @@
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/351>`_.
+.. contents::
+
+Provide an extension to disallow syntax that would desugar to run-time errors without a ``-fdefer-*`` flag.
+
+Motivation
+----------
+
+Going back to Milner's "Well-Typed expressions do not go wrong", typed functional programmers, and Haskellers in particular, usually prefer their programs to be free of run-time failures stemming from programmer error.
+For those learning Haskell, learning this mindset of being fastidious and aiming for code that is correct-by-construction about even banal errors is sometimes as important and novel as learning the language itself.
+We should help those learners by trying to guide them in that direction every step of the code they write.
+
+Unfortunately, the compiler and language do not always help with this.
+We have old syntax that can not only introduce run-time errors, but does so implicitly.
+Incomplete pattern matches, incomplete record updates, etc., are all seldom-asked-for "conveniences" that just get in the way of a programmer trying to understand where their program can "go wrong".
+
+I think most of us would agree with the above paragraph.
+But many of us also would say it's a solved problem, because we have ``-Werror=...`` to catch all these things.
+Why write yet another proposal, for yet another language extension, when this is solved problem?
+
+Well, I just don't think ``-Werror=...`` is the right mechanism to rely on.
+This is, of course, a matter of opinion, but to me warnings are for "extra" analyses, not "essential" ones like type checking which enforce the language proper.
+Deciding whether e.g. pattern matching is exhaustive feels to me like one of those essential analyses.
+I am asking that we switch our vantage point and instead of considering programs with implicit partiality well-formed programs that are iffy, consider them ill-formed programs that we can permissively accept anyways.
+
+So, what's the practical upshot of this lawyering?
+And what mechanism might we use rather than ``-Werror=...``?
+There is another mechanism for controlling run-time errors which I think is a perfect model for how these programs should be dealt with: ``-fdefer-type-errors``.
+As I mentioned above, nobody would propose type checking is some extraneous analysis a Haskell implementation need not do by default.
+The "defer" in the name makes clear we are taking what is normally a compile-time error and deferring it to run-time behind additional branching.
+Whether for quick slap-dash programming, or arguing with proponents of untyped languages, ``-fdefer-type-errors`` allows GHC to both express its opinion and yet also not straight-jacket the user into any one policy.
```

Comment:
Without type-checking, a Haskell program doesn't really have any meaning (more formally, it doesn't have a sensible denotational or operational semantics). In the absence of types an expression like `mempty` is simply meaningless;  `-fdefer-type-errors` throws in runtime errors in an implementation-defined way to patch it up. This is very different to (gradually) dynamically typed languages, where untyped expressions have a more-or-less sensible semantics.

In contrast, even with the coverage checker disabled (or buggy), a Haskell program has perfectly well-defined semantics: it throws a runtime error if an uncovered pattern occurs during evaluation. Thus I think that for Haskell, coverage checking really is an "extra" analysis, whereas type-checking is an "essential" one.

One can certainly imagine a language for which coverage checking really is necessary as part of the language specification, but Haskell isn't such a language. Some dependently-typed languages even require totality, which allows optimizing away runtime computations because there is no requirement to check proofs at runtime, but then they have undefined semantics if a non-total program slips past the compiler. 

Sorry to harp on. :wink: I think we are in agreement in preferring a programming style where the absence of incompleteness is obvious to the compiler, with the compiler warning us if it can't understand why a program is complete. I'm just not convinced this motivates a new language extension and a lot of churn around warning names.

Reply:
> In contrast, even with the coverage checker disabled (or buggy), a Haskell program has perfectly well-defined semantics.

We *choose* to give it those semantics. We could also choose not too. I agree the math doesn't force us to exclude such programs, but it doesn't need us to include them either. It's a matter of opinion.

I want there to be a generation of Haskell programs who've never not had incomplete patterns be an error (by default), and who would be shocked to learn the Haskell ever worked differently. If you have an alternative way to get there, I'm all ears.